File: HW/hw2/HW2.py
'''
Created on Feb 2, 2024

@author: alexxavier
'''



class HW2:
    def is_bad_version(self, x):
        if(x == 7 or x==8 or x==9 or x==10):
            return True
        else:
            return False
        
    def index_of_bad_version(self, x):
        start = 0;
        end = len(x)-1
        
        while(start <= end):
            mid = (start+end)//2

            if self.is_bad_version(x[mid]) and (mid == 0 or not self.is_bad_version(x[mid-1])):
                return mid
            elif self.is_bad_version(x[mid]):
                end = mid - 1
            else:
                start = mid + 1
        return -1
             
       
x = HW2()
versions = [1,2,3,4,5,6,7,8,9,10]
print(x.index_of_bad_version(versions))

File: HW/hw2/HW2.java
package HW;

public class HW2 {

	public static boolean isBadVersion(int x) {
		if (x == 8 || x == 9 || x == 10) {
			return true;

		} else {
			return false;
		}

	}

	public int indexOffirstBadVersion(int[] arr) {
		int startIndex = 0;
		int endIndex = arr.length - 1;

		while (startIndex < endIndex) {
			int mid = startIndex + (endIndex - startIndex) / 2;

			if (isBadVersion(arr[mid])) {
				return mid;
			}

			if (mid + 1 < arr.length && !isBadVersion(arr[mid + 1])) {
				startIndex = mid + 1;
			}

			else {
				endIndex = mid - 1;
			}

		}
		return -1;

	}

	public static void main(String[] args) {
		int[] arr = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
		HW2 x = new HW2();
		System.out.println(x.indexOffirstBadVersion(arr));
	}

}
